app/drivers/distributed_manager.go

```diff
@@ -47,6 +94,78 @@ func (d *DistributedManager) CleanPools(ctx context.Context, destroyBusy, destro
 	return returnError
 }
 
+func (d *DistributedManager) cleanPool(ctx context.Context, pool *poolEntry, query *types.QueryParams, destroyBusy, destroyFree bool) error {
+	if !destroyBusy && !destroyFree {
+		return fmt.Errorf("distributed dlite: both destroyBusy and destroyFree cannot be false")
+	}
+
+	// Determine which states are allowed for cleanup
+	var allowedStates []types.InstanceState
+	if destroyBusy {
+		allowedStates = append(allowedStates, types.StateInUse)
+	}
+	if destroyFree {
+		allowedStates = append(allowedStates, types.StateCreated, types.StateHibernating)
+	}
+
+	// Set the pool name on the query parameters
+	query.PoolName = pool.Name
+
+	// 1. Find and claim all matching instances
+	var instancesToDestroy []*types.Instance
+	for {
+		// Claim one instance at a time to avoid long transactions and to process them sequentially.
```

Comment:
we need a trade off here. For large pools, this creates many round-trips to the database and Increases contention when multiple runners are cleaning up simultaneously